admin/used_flusher.go

```diff
@@ -50,39 +40,57 @@ func newUsedFlusher(logger *zap.Logger, db database.DB) *usedFlusher {
 	return used
 }
 
+func (u *usedFlusher) Deployment(id string) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	u.deployments[id] = true
+}
+
+func (u *usedFlusher) Close() {
+	u.cancel()
+	u.flush()
+	u.flushWg.Wait()
+}
+
 func (u *usedFlusher) runBackground() {
-	ticker := time.NewTicker(flushDuration)
+	ticker := time.NewTicker(flushInterval)
 	for {
 		select {
 		case <-ticker.C:
-			u.flush(u.ctx)
+			u.flush()
 		case <-u.ctx.Done():
 			ticker.Stop()
 			return
 		}
 	}
 }
 
-func (u *usedFlusher) flush(ctx context.Context) {
-	u.logger.Info(`flush deployments`, zap.Int("no of deployments", len(u.deployments)), observability.ZapCtx(u.ctx))
+func (u *usedFlusher) flush() {
 	u.flushWg.Add(1)
 	defer u.flushWg.Done()
-	if len(u.deployments) > 0 {
-		u.lock.Lock()
-		deployments := u.deployments
-		u.deployments = make(map[string]bool)
-		u.lock.Unlock()
 
-		deplIds := make([]string, 0, len(deployments))
+	u.mu.Lock()
+	deployments := u.deployments
+	u.deployments = make(map[string]bool)
+	u.mu.Unlock()
```

Comment:
Why we are not checking this inside `len(u.deployments) > 0`? if no deployments i guess we should not lock it every time this runs

Reply:
It would not be faster – you can't call `len(u.deployments)` without locking it first (locks are not just for writes, they're also needed for reads that can happen during a write).